homeassistant/components/litterrobot/vacuum.py

```diff
@@ -1,123 +1,153 @@
 """Support for Litter-Robot "Vacuum"."""
-from pylitterbot import Robot
+from __future__ import annotations
+
+from typing import Any, Callable
+
+from pylitterbot.enums import LitterBoxStatus
+from pylitterbot.robot import VALID_WAIT_TIMES
+import voluptuous as vol
 
 from homeassistant.components.vacuum import (
     STATE_CLEANING,
     STATE_DOCKED,
     STATE_ERROR,
-    SUPPORT_SEND_COMMAND,
+    STATE_PAUSED,
     SUPPORT_START,
     SUPPORT_STATE,
     SUPPORT_STATUS,
     SUPPORT_TURN_OFF,
     SUPPORT_TURN_ON,
     VacuumEntity,
 )
+from homeassistant.config_entries import ConfigEntry
 from homeassistant.const import STATE_OFF
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import config_validation as cv, entity_platform
+from homeassistant.helpers.entity import Entity
 
 from .const import DOMAIN
-from .hub import LitterRobotEntity
+from .entity import LitterRobotControlEntity
+from .hub import LitterRobotHub
 
 SUPPORT_LITTERROBOT = (
-    SUPPORT_SEND_COMMAND
-    | SUPPORT_START
-    | SUPPORT_STATE
-    | SUPPORT_STATUS
-    | SUPPORT_TURN_OFF
-    | SUPPORT_TURN_ON
+    SUPPORT_START | SUPPORT_STATE | SUPPORT_STATUS | SUPPORT_TURN_OFF | SUPPORT_TURN_ON
 )
 TYPE_LITTER_BOX = "Litter Box"
 
+SERVICE_RESET_WASTE_DRAWER = "reset_waste_drawer"
+SERVICE_SET_SLEEP_MODE = "set_sleep_mode"
+SERVICE_SET_WAIT_TIME = "set_wait_time"
 
-async def async_setup_entry(hass, config_entry, async_add_entities):
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: Callable[[list[Entity], bool], None],
+) -> None:
     """Set up Litter-Robot cleaner using config entry."""
-    hub = hass.data[DOMAIN][config_entry.entry_id]
+    hub: LitterRobotHub = hass.data[DOMAIN][entry.entry_id]
 
     entities = []
     for robot in hub.account.robots:
-        entities.append(LitterRobotCleaner(robot, TYPE_LITTER_BOX, hub))
+        entities.append(
+            LitterRobotCleaner(robot=robot, entity_type=TYPE_LITTER_BOX, hub=hub)
+        )
 
     if entities:
```

Comment:
```suggestion
    if not entities:
        return
```

Then outdent `async_add_entities`

Let's not register the entity service if there are no entities that can use it.

Reply:
Actually can this even happen?

Reply:
Theoretically, if someone created an account and didn't finish adding a unit to their account, it could happen.

Reply:
Ok then let's return even earlier if `hub.account.robots` is empty

Reply:
What's the appropriate way to check that a platform wasn't loaded in tests?

Reply:
`vacuum.litterrobot` would be missing from `hass.config.components`

I'm not sure we should check that though as that implementation detail could change in the future.

Probably check to see if the service is registered instead

Reply:
Or should I move the check for `hub.account.robots` to __init__.py and `return False` in `async_setup_entry`?

Reply:
All of the platforms are directly related with a robot at the moment, so it wouldn't make sense to include any if there aren't any robots. Maybe something like:
```python
if hub.account.robots:
    for platform in PLATFORMS:
        hass.async_create_task(
            hass.config_entries.async_forward_entry_setup(entry, platform)
        )
```